Inplace date range picker

NobodysNightmare opened the pull request:
## OpenProject work package

http://community.openproject.org/work_packages/18939
## Notes

This PR is a re-issue of https://github.com/opf/openproject/pull/2662 that has the accidential commit of the `Gemfile.lock` removed.

Due to time constraints the following issues are not handled in this PR, but get separate work packages:
- [date range picker is closed on scrolling](https://community.openproject.org/work_packages/19612)
- [clicking `Clear` does not close the date range picker](https://community.openproject.org/work_packages/19613)
- [it is not possible to edit date custom fields via inplace edit](https://community.openproject.org/work_packages/19614)

Commit:
Added date range picker on details pane.

Commit:
Fix updating of date range in initial spans.

Commit:
Fix missed semicolon and line length.

Commit:
Add of manual editing of date in input.

Commit:
Add semicolon to the end of line.

Commit:
Add tests for date range picker.

Commit:
Fix date range picker spec.

Commit:
Fix tests.

Commit:
Fix positioning in Firefox

Commit:
Fix date range picker.

Add button for clearing dates. Now is possible to set one date.
Add week days in calendar and the starting of week is possible to set
from settings.

Commit:
Changes to range datepicker.

Commit:
Refactor of date range picker added new use cases of it.

Commit:
Add ignore file for hound.

Commit:
Fix hound issues.

Commit:
Fix tests and z-index of datepicker.

Commit:
Fix showing of date in edit mode.

Commit:
Rebased on @0xF013's branch

Commit:
Fix baseline, scrolling for datepicker and tests.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-inplace-date-range-picker.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-inplace-date-range-picker.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
Protractor tests failed two times in a row. If they don't turn green now (third try), I'll consider them as correctly failing...

floriank commented:
> ERROR - Unable to start a WebDriver session.

Does not seem like a test issue per se. @myabc any ideas ?

myabc commented on frontend/tests/integration/protractor.conf.js:
```diff
@@ -32,7 +32,7 @@ exports.config = {
   // Capabilities to be passed to the webdriver instance.
   capabilities: {
-    'browserName': 'firefox'
+    'browserName': 'chrome'
```
@hirtie-maxim this shouldn't be changed.

myabc commented:
@floriank 

> Does not seem like a test issue per se. @myabc any ideas ?

See https://github.com/opf/openproject/pull/2817/files#r28232028

Travis doesn't support Chrome.

floriank commented:
@NobodysNightmare would you be so kind and revert the `protractor.conf` ?

nvm. I noticed you just added me as collab.

Commit:
revert browser capability

floriank commented on frontend/app/config/configuration-service.js:
```diff
@@ -90,5 +90,11 @@ module.exports = function() {
       return this.settings.enabled_modules.indexOf(module) >= 0;
     },
+    startOfWeekPresent: function() {
+      return this.displaySettingPresent('start_of_week') && this.setting.display.start_of_week != '';
```
this was introduced by @hschink intiall, nothing @NobodysNightmare , @hirtie-maxim or myself touched.

NobodysNightmare replied:
Lets wait if he returns to fix that styling issue xD

NobodysNightmare commented on .jshintignore:
```diff
@@ -4,3 +4,5 @@ app/assets/javascripts/jstoolbar/**/*
 app/assets/javascripts/project/**/*
 app/assets/javascripts/lib/**/*
+frontend/tests/*
+frontend/tests/**/*
```
Is this something we really want?

Isn't JSHint something like rubocop?

myabc replied:
@NobodysNightmare right. We shouldn't be ignoring tests.

floriank replied:
It is - this should probably also be reverted. However, I do remember, that jshint complaints are a thing with the tests, so therefore we could leave it be.

myabc replied:
@floriank then we should adjust `.jshintrc` or add a special comment inline (e.g. `/*jshint expr: true*/`)

floriank replied:
I'll revert it, let's see what happens.

NobodysNightmare commented:
@myabc @floriank Do you see any other file changes that should probably not be part of this PR?

I don't know all the configuration files one might accidentially change as frontend developer ^^ (similar to `Gemfile.lock`, `production.rb`, etc.)

myabc commented on frontend/app/services/timezone-service.js:
```diff
@@ -1,30 +1,3 @@
-//-- copyright
```
@hirtie-maxim why did you delete the copyright notice?

floriank commented on frontend/app/config/configuration-service.js:
```diff
@@ -90,5 +90,11 @@ module.exports = function() {
+    }
   };
-};
+};
```
Files should end in a newline

floriank commented on frontend/app/services/timezone-service.js:
```diff
@@ -52,8 +25,18 @@ module.exports = function(ConfigurationService, I18n) {
 
   return TimezoneService;
-};
+};
```
hhhnnhnhnhgrl.

floriank commented on frontend/app/ui_components/date/date-range-picker-directive.js:
```diff
@@ -0,0 +1,226 @@
+      });
+    }
+  };
+};
```
same here.

floriank commented on frontend/public/templates/components/inplace_editor/date/date_range_picker.html:
```diff
@@ -0,0 +1,3 @@
+<div class="date-range-picker">
+	<input ng-model="editableStartDate" type="text" class="start" /><span>-</span><input ng-model="editableEndDate" type="text" class="end" />
+</div>
```
http://stackoverflow.com/questions/729692/why-should-files-end-with-a-newline

myabc commented on app/assets/javascripts/application.js.erb:
```diff
@@ -77,6 +77,34 @@ jQuery(document).ready(function ($) {
     return false;
   });
 
+  $.extend($.datepicker, {
```
ARGHH!! @hirtie-maxim please don't add anything new to `application.js.erb`. Adding new scripts here should be considered deprecated. `frontend/app/misc/datepicker-defaults.js` is probably the right file to append to.

myabc commented on frontend/app/templates/components/inplace_editor/date/date_range_picker.html:
```diff
@@ -0,0 +1,3 @@
+<div class="date-range-picker">
+	<input type="text" class="start" /><span>-</span><input type="text" class="end" />
```
@hirtie-maxim spaces not tabs please!

myabc commented on frontend/public/templates/components/inplace_editor/date/date_range_picker.html:
```diff
@@ -0,0 +1,3 @@
+<div class="date-range-picker">
+	<input ng-model="editableStartDate" type="text" class="start" /><span>-</span><input ng-model="editableEndDate" type="text" class="end" />
```
@hirtie-maxim spaces not tabs please!

Commit:
readd the copyright notice in the timzone service

Commit:
revert jshintignore

Commit:
fix code style

includes tabs to spaces as well as file endings

Commit:
move datepicker extension

remove the datepciker extension from application.js.erb

Commit:
remove logging from test

Commit:
xit a broken test

this xits a tests which has (most likely) been broken by the datepicker. It
could be a simple selection issue, but it will have to wait (no functionality
broken)

Commit:
use proper comparison

floriank commented:
This will probably take a little bit.

floriank commented on frontend/app/work_packages/services/work-package-field-service.js:
```diff
@@ -254,6 +263,8 @@ module.exports = function(
         break;
       case 'User':
         displayStrategy = 'user';
```
@hirtie-maxim there is a `break` missing here, leading to a problem where all user fields will be treated as `daterange` inputs.

Commit:
fix type for field in question

This fixes a issue where the daterange type bled into the User type for editable
fields.

Commit:
fix javascript style for datepicker defaults

Commit:
cleanup test object structure

Commit:
clean up lines

floriank commented on frontend/app/misc/datepicker-defaults.js:
```diff
@@ -68,4 +68,33 @@ jQuery(function($) {
+      // Call the original
+      this._inlineDatepicker2(target, inst);
+
+      var beforeShow = $.datepicker._get(inst, 'beforeShow');
```
I have no friggin idea here.

floriank commented:
I cleaned everything up, no need for ignoring the frontend specs. The implementation for the bugs has to wait though, I need my beauty sleep.

NobodysNightmare commented:
@myabc Can you have a look at this?
I don't feel to be able to review the JS changes @floriank made...

floriank commented:
The JS changes is mostly styling, cleaning up lines and the case of the forgotten `break`  by @hirtie-maxim
